src/test/java/com/example/springonlinebookstore/controller/BookControllerTest.java

```diff
@@ -0,0 +1,155 @@
+
+import com.example.springonlinebookstore.dto.book.BookDto;
+import com.example.springonlinebookstore.dto.book.CreateBookRequestDto;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.math.BigDecimal;
+import java.util.List;
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.http.MediaType;
+import org.springframework.security.test.context.support.WithMockUser;
+import org.springframework.test.context.jdbc.Sql;
+import org.springframework.test.context.jdbc.SqlMergeMode;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.MvcResult;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+import org.springframework.web.context.WebApplicationContext;
+
+@Sql(scripts = {"classpath:database/delete-all.sql"},
+        executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@SqlMergeMode(SqlMergeMode.MergeMode.MERGE)
+public class BookControllerTest {
+    protected static MockMvc mockMvc;
+
+    @Autowired
+    private ObjectMapper objectMapper;
+
+    @BeforeAll
+    static void beforeAll(@Autowired WebApplicationContext applicationContext) {
+        mockMvc = MockMvcBuilders
+                .webAppContextSetup(applicationContext)
+                .apply(springSecurity())
+                .build();
+    }
+
+    @Test
+    @DisplayName("Verify save() method works, return dto")
+    @WithMockUser(username = "admin", roles = "ADMIN")
+    @Sql(scripts = {"classpath:database/add-categories-to-database.sql"})
+    public void saveBook_ValidCreateBookRequestDto_Success() throws Exception {
+        CreateBookRequestDto createBookRequestDto = new CreateBookRequestDto();
+        createBookRequestDto.setAuthor("Author");
+        createBookRequestDto.setPrice(BigDecimal.valueOf(1));
+        createBookRequestDto.setCategories(List.of(1L));
+        createBookRequestDto.setTitle("Title");
+        createBookRequestDto.setCoverImage("Image");
+        createBookRequestDto.setDescription("Description");
+        createBookRequestDto.setIsbn("9788869185182");
+
+        String jsonRequest = objectMapper.writeValueAsString(createBookRequestDto);
+
+        MvcResult result = mockMvc.perform(post("/books")
+                        .content(jsonRequest)
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().is(200))
+                .andReturn();
+
+        BookDto actual = objectMapper.readValue(
+                result.getResponse().getContentAsString(), BookDto.class);
+        assertNotNull(actual);
+        EqualsBuilder.reflectionEquals(getBookDto(), actual, "id");
```

Comment:
use `jsonPath(...)` to simplify testing output of the controller. 
```suggestion
.andExpect(jsonPath("$[0].foo", is("bar"))) // example
```

Reply:
refactor all controller tests regarding this comment